cdap-watchdog/src/main/java/co/cask/cdap/logging/clean/FileMetadataScanner.java

```diff
@@ -0,0 +1,236 @@
+
+  private static final byte[] NEW_ROW_KEY_PREFIX = LoggingStoreTableUtil.NEW_FILE_META_ROW_KEY_PREFIX;
+  private static final byte[] NEW_ROW_KEY_PREFIX_END = Bytes.stopKeyForPrefix(NEW_ROW_KEY_PREFIX);
+
+  // cut-off time discount from actual transaction timeout
+  private static final int CUTOFF_DISCOUNT = 5;
+  private final Transactional transactional;
+  private final DatasetManager datasetManager;
+
+  public FileMetadataScanner(DatasetManager datasetManager, Transactional transactional) {
+    this.transactional = transactional;
+    this.datasetManager = datasetManager;
+  }
+
+  /**
+   * scans for meta data in old format which has expired the log retention.
+   */
+  @VisibleForTesting
+  List<byte[]> scanAndDeleteOldMetaData(int transactionTimeout, final int cutoffTransactionTime) {
+    final List<byte[]> deletedEntries = new ArrayList<>();
+    try {
+      transactional.execute(transactionTimeout, new TxRunnable() {
+        public void run(DatasetContext context) throws Exception {
+          Stopwatch stopwatch = new Stopwatch();
+          stopwatch.start();
+          Table table = LoggingStoreTableUtil.getMetadataTable(context, datasetManager);
+          // create range with tillTime as endColumn
+          Scan scan = new Scan(OLD_ROW_KEY_PREFIX, OLD_ROW_KEY_PREFIX_END, null);
+          try (Scanner scanner = table.scan(scan)) {
+            Row row;
+            while ((row = scanner.next()) != null) {
+              if (stopwatch.elapsedTime(TimeUnit.SECONDS) > cutoffTransactionTime) {
+                break;
+              }
+              byte[] rowKey = row.getRow();
+              // delete all columns for this row
+              table.delete(rowKey);
+              deletedEntries.add(rowKey);
+            }
+          }
+        }
+      });
+    } catch (TransactionFailureException e) {
+      LOG.warn("Got Exception while deleting old metadata", e);
+      // not required as we dont delete files in old format, but still safe to clear.
+      deletedEntries.clear();
+    }
+    LOG.info("Deleted {} entries from the meta table for the old log format", deletedEntries.size());
+    return deletedEntries;
+  }
+
+  /**
+   * scans for meta data in old format which has expired the log retention.
+   * @param tillTime time till which files will be deleted
+   * @param transactionTimeout transaction timeout to use for scanning entries, deleting entries.
+   * @return list of DeleteEntry - used to get files to delete for which metadata has already been deleted
+   */
+  public List<DeleteEntry> scanAndGetFilesToDelete(final long tillTime, final int transactionTimeout) {
+    final List<DeleteEntry> toDelete = new ArrayList<>();
+    if (transactionTimeout < CUTOFF_DISCOUNT) {
+      LOG.warn("Transaction timeout for log cleanup is really low {}s", transactionTimeout);
+    }
+    final int cutOffTransactionTime =
+      transactionTimeout > CUTOFF_DISCOUNT ? transactionTimeout - CUTOFF_DISCOUNT : transactionTimeout;
```

Comment:
If timeout is smaller than cutoff, we still shouldn't run till the tx timeout. Maybe just divide the tx timeout by half.